### PR TITLE
Add filter

### LINKED ISSUE #79: Add a mild word filter
Add a `wordFilter` default function for the bot to run on messages if they are not commands, then (should it be necessary) delete the message and warn the user. 

* A list of words deemed non-acceptable by staff will need to be made for this.
* Currently an extra `else if` would nee to be added in app.js to call the function since it wouldn't start with cc!:
```
client.on('message', msg => {
  if (msg.content.substring(0, 3) === 'cc!' && !(msg.member === client)) {
    commandParser(msg);
  } else if (!(msg.member === client) {
    // call to function
  }
});
```

On hold until the warning/infractions systems are ready.

### PR DESCRIPTION
**Pull request by Shaylin-TRK**
## What issue is this solving?

Closes #79 

### Description

Adds a mild word filter to the bot to help with moderation. Currently the list of offensive words is rather small but can easily be expanded to include more.

The bot currently takes action if it finds 3+ of the lesser words listed in a message or a single of the stronger words. Action taken is to delete the message inform staff in audit logs that action was taken, add a note to the database for future reference, and finally DM the user asking them to refrain from using offensive content and giving them a copy of their message to make a rewrite easier.

## Any helpful knowledge/context for the reviewer?

- Is a re-seeding of the database necessary? NO
- Any new dependencies to install? NO
- Any special requirements to test? YES
  You cannot have SU, Mod, or Admin roles for the bot to filter your message

### Please make sure you've attempted to meet the following coding standards
- [x] Code has been tested and does not produce errors
- [x] Code is readable and formatted
- [x] There isn't any unnecessary commented-out code

### CODE REVIEW

**Review by Vic-ST — CHANGES_REQUESTED** [review]
Looks great to me, seems to run well, one thing, can we add a couple more words to that list before merging?

**Comment by Shaylin-TRK** [comment]
Just added a bunch more. I intentionally omitted some words including things like ass and dick because they are both used outside of as hurtful words. If you think they should be included anyway or there are more that should be added please let me know.

**Review by Vic-ST — APPROVED** [review]
Looks great!

File: commands/moderation/filter.js
Comment (aedwardg) [inline commands/moderation/filter.js:64]: I don't think we really need the punctuation at the end (`[.,/\\:;\*\?\!\@\#\$]?`) since `RegExp.test()` will still match the word whether it's got punctuation at the end or not.

Let's remove it unless there's a specific reason we need it.

File: commands/moderation/filter.js
Comment (aedwardg) [inline commands/moderation/filter.js:55]: I'm starting to think we should separate out a version or two of @fedGL's `isHighRoller` function into its own module so that we can just import it and use it where we need.  Right now I think something similar to this code is defined in almost every command file, which is the opposite of the DRY principle...

That should probably be done as a separate refactoring issue.

File: commands/moderation/filter.js
Comment (aedwardg) [inline commands/moderation/filter.js:113]: I like that this is separated out into its own function.  I think most of the rest of `execute` could also be refactored into functions to make it easier to glance at and see the gist of what is happening.  For example, there could be a `swearCounter` function and a `addNoteAndLog` function, or whatever names you feel work.

This isn't a must for this PR - it could be done in a separate one if you want.

File: commands/moderation/filter.js
Comment (aedwardg) [inline commands/moderation/filter.js:61]: Since we may change how we get all these words in the future, I won't ask you to make this change yet, but I thought it would be worth noting that this RegExp could be shortened significantly by not using the full word for each variation.
  
For example, this would cover 12 of the included words, plus some variations that use two "m"s instead of "mn":
```javascript
/((god)?dam[mn](ed|it)?)|((mother)?fuck(ed|er|ing|wad)?)/
```

**Review by aedwardg — CHANGES_REQUESTED** [review]
Great job @Shaylin-TRK, works as intended! 

I just have one requested change (removing the punctuation) along with a handful of comments.

If you feel like further refactoring the `execute` logic into smaller functions in this PR that would be great, but I'll still approve if you'd rather save that for another PR.

File: commands/moderation/filter.js
Comment (Shaylin-TRK) [reply commands/moderation/filter.js]: Did not know this! Done

File: commands/moderation/filter.js
Comment (Shaylin-TRK) [reply commands/moderation/filter.js]: I went ahead and made a function like this in the filter.js file so it will be easier to refactor in the future.

File: commands/moderation/filter.js
Comment (Shaylin-TRK) [reply commands/moderation/filter.js]: Went ahead and split everything up

File: commands/moderation/filter.js
Comment (Shaylin-TRK) [reply commands/moderation/filter.js]: I figured it could be done a lot shorter but since we weren't certain on the word list yet, I let it long so it would be easier for me to modify. Noted for the future though.

**Review by aedwardg — COMMENTED** [review]
@Shaylin-TRK, this looks pretty good, but after the meeting last week I think the group wants to go in a slightly different direction for the word filter.

Rather than auto-moderating most of these words, we think we can probably drop the `lightWords` list completely and then only auto-moderate the sexually-explicit/racist words.

The rest of the `strongWords` can instead just be logged into a separate channel specifically for manual language review (maybe #swear-jar or something).  That way, we don't go overboard and lose users, but we also get notified any time there is a serious swear word so that we can review on a case-by-case basis.

@CodecademyCommunity/discord-bot let me know if I missed anything!

**Comment by Vic-ST** [comment]
> The rest of the `strongWords` can instead just be logged into a separate channel specifically for manual language review (maybe #swear-jar or something). That way, we don't go overboard and lose users, but we also get notified any time there is a serious swear word so that we can review on a case-by-case basis.

I'll be creating this channel now on the main server and on the dev server. Might want to add a check if the channel isn't there then create it though...

**Comment by Shaylin-TRK** [comment]
My apologies for the wait. New word filtering criteria is now finished. I've decided to hold off for now on having it create a channel if swear-jar is missing, but it checks for the channel before attempting to send there to keep it from crashing. It is also available to build off of these `if` blocks in the future.

**Review by aedwardg — COMMENTED** [review]
Great work @Shaylin-TRK, thanks for doing this!

Just format and lint this PR and it'll be ready to approve.

**Comment by Shaylin-TRK** [comment]
Done, formatted and linted.

**Review by aedwardg — APPROVED** [review]
Looks good to me!
 
If there are any more edge cases or if we don't like how the bot handles a certain word, we can always update this.

I'm excited to have the swear words logged in #swear-jar, so it'll make it easier to find them in channels where there's a lot of activity or excessive swearing that occurred while I was AFK.

Great job.

**Pull request merged**
